File: code/c/count-frames.c
#include <stdio.h>
#include <stdlib.h>

#define BUFFER_SIZE 1024
#define MARKER_HIGH 0xFF
#define MARKER_LOW  0xFD

int count_frames_in_xtxt(const char *file_path) {
    FILE *file = fopen(file_path, "rb");
    if (file == NULL) {
        perror("Error opening file");
        return -1;
    }

    unsigned char buffer[BUFFER_SIZE];
    int frame_count = 0;
    size_t bytes_read;
    unsigned char prev_byte = 0;

    while ((bytes_read = fread(buffer, 1, BUFFER_SIZE, file)) > 0) {
        for (size_t i = 0; i < bytes_read; i++) {
            unsigned char current_byte = buffer[i];
            if (prev_byte == MARKER_HIGH && current_byte == MARKER_LOW) {
                frame_count++;
            }
            prev_byte = current_byte;
        }
    }

    if (ferror(file)) {
        perror("Error reading file");
        fclose(file);
        return -1;
    }

    fclose(file);
    return frame_count;
}

int main() {
    const char *file_path = "input.xtxt";

    int frame_count = count_frames_in_xtxt(file_path);
    if (frame_count >= 0) {
        printf("Total frames: %d\n", frame_count);
    } else {
        printf("Failed to count frames.\n");
    }

    return 0;
}

File: code/py/count-frames.py
def count_frames_in_xtxt(file_path):
    """
    Counts the number of frames in an XTXT file by detecting the
    Next Frame Marker (NFM), defined as the sequence 0xFF 0xFD.

    :param file_path: Path to the .xtxt file
    :return: Number of frames detected
    """
    NFM = bytes([0xFF, 0xFD])  # Define the Next Frame Marker
    BUFFER_SIZE = 1024         # Process the file in 1 KB chunks

    try:
        frame_count = 0
        prev_byte = None

        with open(file_path, 'rb') as file:
            while chunk := file.read(BUFFER_SIZE):
                for i in range(len(chunk)):
                    current_byte = chunk[i]

                    # Check for boundary-spanning marker
                    if prev_byte == NFM[0] and current_byte == NFM[1]:
                        frame_count += 1

                    prev_byte = current_byte

        return frame_count

    except FileNotFoundError:
        print(f"Error: File '{file_path}' not found.")
        return -1
    except Exception as e:
        print(f"An error occurred: {e}")
        return -1

if __name__ == "__main__":
    # Specify the path to the .xtxt file
    file_path = "input.xtxt"

    # Count the frames and print the result
    frames = count_frames_in_xtxt(file_path)
    if frames >= 0:
        print(f"Total frames: {frames}")
